rfc/graph-feature-flags.md

```diff
@@ -0,0 +1,242 @@
+```
+
+This will publish the new Subgraph Schema, assuming that it composes with the existing Supergraph.
+If it doesn't compose, the request will fail, and the Subgraph will not be published.
+If future changes to other Subgraphs are incompatible with the feature flagged Subgraph,
+the Subgraph will be disabled until the composition checks pass again.
+
+This option is only advisable for staging environments and should not be used in production.
+
+How to enable a feature flag and gradually roll out traffic to the new Subgraph is up to the client:
+
+Add a Header Rule your Edge Gateway that sets the `X-WG-Feature-Flag` Header for 1% of your traffic.
+
+### Monitoring Gradual Feature Rollout
+
+A key aspect of rolling out new features gradually is monitoring.
+E.g. when we move a field from a monolith to a new Subgraph,
+we want to monitor the performance, latency, correctness, and error rate of the new "feature" compared to the existing solution.
+If a problem arises, we want to be able to immediately disable the feature flag.
+If everything goes well, we want to gradually increase the percentage of traffic that is sent to the new Subgraph.
+
+To achieve this, we need to tag metrics and traces with the feature flags that were enabled for a specific request,
+and we need to index these metrics and traces in a way that allows us to query them efficiently.
+
+Once we have this data, we can add a new section in Cosmo Studio that shows all feature flags that are currently enabled,
+and how they compare against the "main" Graph in terms of performance, correctness, and error rate.
+
+### Improving the Developer Experience of Gradual Feature Rollout with Automation
+
+We can further improve the developer experience of rolling out new features gradually by automating the process.
+In the "Feature Flags" section of Cosmo Studio, we can add a function that will automatically roll out a new feature in steps from
+1%-10% in 1% increments, then 10%-30% in 5% increments, and finally 30%-100% in 10% increments.
+
+This function will increase the percentage of traffic that is sent to the new Subgraph based on the performance, correctness, and error rate of the new "feature" compared to the existing solution. If a threshold is exceeded, the function will stop the rollout and disable the feature flag.
+
+What this would allow teams to do is to set up a new feature flag,
+and then let the system automatically roll out the feature without requiring the platform team to deploy extra infrastructure or handle the rollout manually.
+
+The goal here is to avoid the platform team becoming a bottleneck for rolling out new features,
+but instead, to empower the teams to roll out new features safely and quickly,
+and to provide them but also the platform team with the necessary tools to monitor and observe rollouts.
+
+As you scale your API across more and more teams,
+we want to make sure that the platform team can focus on the platform itself,
```

Comment:
The more feature flags a subgraph has, the more variants it gets. Also here tooling is the key. I also want to understand what services "infrastructure" needs to run to serve all or X variants.